drivers/epd_bw_spi/epd_bw_spi.c

```diff
@@ -0,0 +1,321 @@
+{
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_WRITE_RAM, NULL, 0);
+}
+
+void epd_bw_spi_clear(epd_bw_spi_t *dev)
+{
+    epd_bw_spi_fill(dev, 0, dev->size_x, 0, dev->size_y,
+                    EPD_BW_SPI_COLOR_WHITE);
+}
+
+void epd_bw_spi_fill(epd_bw_spi_t *dev, uint8_t x1, uint8_t x2, uint16_t y1,
+                     uint16_t y2, uint8_t color)
+{
+    epd_bw_spi_set_area(dev, x1, x2, y1, y2);
+
+    spi_acquire(dev->params.spi, dev->params.cs_pin, SPI_MODE_0,
+                dev->params.spi_clk);
+    epd_bw_spi_cmd_start(&dev->params, EPD_BW_SPI_CMD_WRITE_RAM, true);
+
+    uint16_t size = ((x2 - x1) >> 3) * (y2 - y1);
+    for (uint16_t i = 0; i < size - 1; i++) {
+        spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true, color);
+    }
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, false, color);
+
+    spi_release(dev->params.spi);
+}
+
+void epd_bw_spi_fill_pixels(epd_bw_spi_t *dev, uint8_t x1, uint8_t x2,
+                            uint16_t y1, uint16_t y2,
+                            uint8_t *px)
+{
+    epd_bw_spi_set_area(dev, x1, x2, y1, y2);
+    epd_bw_spi_write_buffer(dev, px, (x2 - x1) * (y2 - y1));
+}
+
+void epd_bw_spi_set_area(epd_bw_spi_t *dev, uint8_t x1, uint8_t x2, uint16_t y1,
+                         uint16_t y2)
+{
+    /* Set X bounds */
+    uint8_t x_data[] = {
+        (x1 >> 3) & 0x1F,
+        ((x2 - 1) >> 3) & 0x1F,
+    };
+
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_SET_RAM_X, x_data,
+                         sizeof x_data);
+
+    /* Set Y bounds */
+    le_uint16_t y_data[2] = { 0 };
+    uint8_t y_size;
+    if (dev->controller.size_y <= 255) {
+        y_data[0].u8[0] = y1;
+        y_data[0].u8[1] = y2 - 1;
+        y_size = 2;
+    }
+    else {
+        y_data[0] = byteorder_btols(byteorder_htons(y1 & 0x01FF));
+        y_data[1] = byteorder_btols(byteorder_htons((y2 - 1) & 0x01FF));
+        y_size = 4;
+    }
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_SET_RAM_Y,
+                         (uint8_t *)y_data, y_size);
+
+    /* Set counters to start positions */
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_SET_RAM_X_ADDR_COUNTER,
+                         x_data, 1);
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_SET_RAM_Y_ADDR_COUNTER,
+                         (uint8_t *)y_data, 2);
+}
+
+void epd_bw_spi_write_buffer(epd_bw_spi_t *dev, const uint8_t *buf, size_t len)
+{
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_WRITE_RAM, buf, len);
+}
+
+void epd_bw_spi_sleep(epd_bw_spi_t *dev)
+{
+    uint8_t data[] = { 0x01 };
+
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_DEEP_SLEEP_MODE, data, 1);
+}
+
+void epd_bw_spi_wake(epd_bw_spi_t *dev)
+{
+    /* Give a low pulse on the reset pin */
+    if (dev->params.rst_pin != GPIO_UNDEF) {
+        gpio_clear(dev->params.rst_pin);
+        xtimer_usleep(EPD_BW_SPI_WAIT_RESET);
+        gpio_set(dev->params.rst_pin);
+    }
```

Comment:
What about making this a function akin to `epd_bw_spi_swreset` like:

```c
void epd_bw_spi_hwreset(epd_bw_spi_t *dev)
{
    /* Give a low pulse on the reset pin */
    if (dev->params.rst_pin != GPIO_UNDEF) {
        gpio_clear(dev->params.rst_pin);
        xtimer_usleep(EPD_BW_SPI_WAIT_RESET);
        gpio_set(dev->params.rst_pin);
        xtimer_usleep(EPD_BW_SPI_WAIT_RESET);
    }
}
```

With an additional 1 ms wait time used/needed for SSD1608 IC drivers, this way to reset the device at `epd_bw_spi_init_[full/part]` (as done for SSD1608) and should do no harm for other ICs that don't have a reset pin nor are required to be reset at initialization.

Reply:
_Btw digging up in the deepest of the internet I've found a [datasheet for SSD1608](https://annejan.com/media/SSD1608_1.2.pdf) used on Waveshare 1.54 in e-Paper devices_

Reply:
Thanks, changed!

Also: thanks for the datasheet!